### ISSUE TITLE
(Re-)define CI process flow

### ISSUE THREAD

**Issue by baentsch** [open-quantum-safe/liboqs#907]
This issue is to discuss, formulate and agree, initially the need for and eventually the CI strategy going forward (and then implement it).

As already discussed, there are different types of (CI) tests:
- quick sanity checks (such as styletest)
- baseline build tests (with different build-options)
- quick cross platform tests
- deeper tests validating all algorithms are operating OK
- long-running tests like address-sanitizer or timecop

Based on this a strawman proposal:
- Classify all tests we have into one of these categories: must pass 1) during PR, 2) for audit, 3) for main merge, 4) for release
- Build out the pipeline accordingly, maybe using github actions using dedicated compute resources for 2+4 
- Tests passing for 3 should also trigger downstream integration tests
- 1,2,3 would be prerequisite for 4 but 4 could mandate additional tests (we may not even have yet).
- 1 should have a (normal average) time-to-completion target time of 15 minutes; 2 can be arbitrary long if it's gating nothing but 4; 3 should deliver pass/fail within 1 hour. 
- Tests for 2 surely would not have to run cross-platform (though it might be nice)

Rationale: We should be able to make progress resolving issues (at their respective level of complexity: Simple ones should be CI'd&closed fast; downstream problems detected within reasonable time (but automatically); "deep" algorithm problems should be unearthed and resolved in a separate time frame).

Alternative proposals welcome. Proposals for classifying the tests as well as how to implement them, too.

@dstebila @xvzcf @jschanck @christianpaquin (any one else not party to the conversation up to now as well, of course) please click "+1" if you agree we should tackle this problem and if you have more time, additional feedback/suggestions would be great.

**Comment by dstebila** [comment]
I've been discussing with @thomwiggers about options.  He uses Github Actions for PQClean and says that he hasn't run into any compute time limits yet, just concurrency limits (well, and a limit on the number of API calls, but he's got a crazy number (1000+) of jobs).  Github Actions would have the benefit of being able to do self-hosted runners in the future if needed (though there are some security risks on doing self-hosted runners with public repositories that would need to be resolved first).

**Comment by jschanck** [comment]
Github actions are a very good idea, ultimately.

In the (very) short term there's a much lighter weight solution that might work for us. Let's look at some recent job times:
```
4115s		armhf
3810s		arm64
2884s		armel
1254s		address-sanitizer
544s		centos-8
537s		alpine-noopenssl
491s		ubuntu-focal-noopenssl
485s		debian-buster
480s		alpine
449s		ubuntu-focal-clang9
447s		macOS-shared
436s		ubuntu-focal-shared-noopenssl
414s		macOS-noopenssl
58s		ubuntu-focal-buildcheck
49s		stylecheck
```

The arm builds are the biggest problem.

CircleCI has a manual approval process, which we can insert anywhere we like in our build pipeline. So I suggest:
* Add a buildcheck as in #903 (see ubuntu-focal-buildcheck above, this takes about a minute).
* Add "require_buildcheck" to all of the linux_x64 jobs. (so we'll catch broken core components after 1 minute instead of 10.)
* Add a "request-approval" job (https://circleci.com/blog/adding-approval-jobs-to-your-ci-pipeline/) after the linux_x64 jobs but before the arm_emulated jobs.
* Make the arm_emulated jobs depend on request-approval.
* Make address-sanitizer run only on audit branches.

Whoever reviews a PR can approve the arm builds once everything else is passing. This should get our CI iterations down to <10 minutes.

**Comment by dstebila** [comment]
That sounds like a great plan, John.  I didn't know about the approval jobs feature.  Can you tell who will be able to approve? I.e., do you think it will detect team members from the Github repo configuration and use those?

**Comment by baentsch** [comment]
Another challenge just observed: We presently trigger all downstream projects' builds on merge, taking all `open-quantum-safe` CI compute resources for quite some time as those builds apparently don't run in parallel. Should we re-consider this automation, too? Particularly `oqs-demos` and `profiling` take _a lot_ of (compute) time (courtesy docker builds of many versions) -- but already oqs-openssl has very extensive tests we may want to whittle down, too, maybe only run on explicit manual approval as per @jschanck's suggestion above. 

`oqs-demos` and `profiling` arguably don't have to be built automatically and could "make do" with a "detached" or nightly build only. Until we have a better idea I'd suggest  removing them from openssl's downstream trigger (initially reverting https://github.com/open-quantum-safe/openssl/pull/272 -- which might be the culprit of this "CI squeeze"). @xvzcf if I recall correct you know of a way to simply remove some access rights and those triggers would not do anything?

**Comment by jschanck** [comment]
> Can you tell who will be able to approve? I.e., do you think it will detect team members from the Github repo configuration and use those?

I tacked my proposal onto #903, it looks like any team member can approve, but someone should check.

**Comment by jschanck** [comment]
Also: can we whittle down the linux_x64 jobs? What are we testing for on alpine, centos, and debian that isn't covered by ubuntu?

**Comment by thomwiggers** [comment]
`alpine` has a different `libc`, namely `musl`, which might make it semi-useful. centos and debian will probably only differ in package versions and compiler versions.

**Comment by xvzcf** [comment]
At one point I was considering trying to set up an as-minimal-as-possible config to test across different libcs and different architectures in the manner of (Wireguard)[https://www.wireguard.com/build-status/], but didn't have the time to follow-through.

**Comment by xvzcf** [comment]
> @xvzcf if I recall correct you know of a way to simply remove some access rights and those triggers would not do anything? Otherwise, please consider [open-quantum-safe/openssl#280](https://github.com/open-quantum-safe/openssl/pull/280)

I think just removing the triggers in the first place is better. We could also just run these triggers as nightly jobs.

**Comment by baentsch** [comment]
>  the best would be if the triggers just scheduled nightly jobs

https://circleci.com/docs/2.0/workflows/#nightly-example

We had that at some point in time --even guarded to not run if nothing had been checked in during the past 24h. Who recalls why we stopped using that?

**Comment by xvzcf** [comment]
> We had that at some point in time --even guarded to not run if nothing had been checked in during the past 24h. Who recalls why we stopped using that?

Those nightly jobs in the downstream projects were set up so as to look for new commits in liboqs, but as a consequence, would fail one day and pass after that. There was also the feedback delay: a commit merged into liboqs wouldn't get fully tested until that night.

What I meant was scheduling a nightly job if and only if a `new_commit` API trigger was made.

**Comment by baentsch** [comment]
> alpine has a different libc, namely musl, which might make it semi-useful. centos and debian will probably only differ in package versions and compiler versions.

+1. In addition alpine is the basis for our docker distributions, so I'd vote for keeping that. With debian I never saw different test results compared to ubuntu, so I'd vote for dropping it. Centos had deviations once or twice. Whether that's enough to keep it within CI? Why not drop it, too?

**Comment by dstebila** [comment]
Given that RedHat terminated CentOS development in December 2020, I'm not inclined to continue with CentOS support unless we hear a clear demand for it.

**Comment by baentsch** [comment]
> What I meant was scheduling a nightly job if and only if a new_commit API trigger was made.

As I have no idea how to do that I went hunting for information. Two ideas now:
1) Create a CCI artifact to store the information that (which) tasks need to be run during the night. Didn't find a way yet how to programatically retrieve "the latest" artifact, though: Only found [this information how to get a specific build's artifacts](https://circleci.com/docs/2.0/artifacts/#downloading-all-artifacts-for-a-build-on-circleci). But the cronjob can't know the number of the "latest build", so I don't see a way forward here.
2) Alternatively, store the same "run-later" information, say via wget, on a server of our choice and trigger nightly CCI tasks from there (via curl task-trigger): That should be doable pretty much straight away.

Other ideas?

**Comment by baentsch** [comment]
> Given that RedHat terminated CentOS development in December 2020, I'm not inclined to continue with CentOS support unless we hear a clear demand for it.

Good point that made me also streamline `open-quantum-safe/ci-containers`: Can you and/or @xvzcf please take a look? Optimizing workflow a bit, I didn't do a PR but placed this in a new branch `main`: If OK, please set that as the new default branch along the lines of #795 so we can dump `master`: Thanks in advance.

**Comment by dstebila** [comment]
Changes in ci-containers look good to me. I changed the default branch to main.

**Comment by jschanck** [comment]
A couple of suggestions:
1) Disable [armel](https://wiki.debian.org/ArmEabiPort) tests. Right now we interpret "armel" as ARM32v7 without hardware floating point. As far as I can tell this is a rare configuration. It would make sense to keep armel if we wanted to cover some older (pre v7) processors. If we want to keep armel we should choose a particular processor to target.
2) Disable test_mem. It's useful for profiling, but we don't collect any of the information that it produces in CI. It tests the same code paths as test_cmdline.

**Comment by dstebila** [comment]
> 1. Disable [armel](https://wiki.debian.org/ArmEabiPort) tests. Right now we interpret "armel" as ARM32v7 without hardware floating point. As far as I can tell this is a rare configuration. It would make sense to keep armel if we wanted to cover some older (pre v7) processors. If we want to keep armel we should choose a particular processor to target.

Fine with me.

> 2. Disable test_mem. It's useful for profiling, but we don't collect any of the information that it produces in CI. It tests the same code paths as test_cmdline.

I'd like to have at least one test path that exercises test\_mem for the purposes of making sure that the test\_mem code itself continues to work, but am fine with removing the bulk of its usage if it slows things down substantially.

**Comment by baentsch** [comment]
[Fulll CI runs in 24mins now](https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs). Much better than when we created the issue. Thanks for all contributions. Case closed.
